bip-0324.mediawiki

```diff
@@ -0,0 +1,467 @@
+ | x, X         := SECP256k1_KEYGEN()                                                   |
+ | CLIENT_HDATA := X                                                                    |
+ |                                                                                      |
+ |               --- CLIENT_HDATA --->                                                  |
+ |                                                                                      |
+ |                                       y, Y           := SECP256k1_KEYGEN()           |
+ |                                       ECDH_KEY       := SECP256k1_ECDH(X,y)          |
+ |                                       SERVER_HDATA   := Y                            |
+ |                                                                                      |
+ |               <-- SERVER_HDATA ----                                                  |
+ |                                                                                      |
+ | ECDH_KEY     := SECP256k1_ECDH(x,Y)                                                  |
+ ----------------------------------------------------------------------------------------
+</pre>
+
+To request encrypted communication (only possible if yet no other messages have been sent or received), the initiating peer generates an EC secp256k1 ephemeral key and sends the corresponding 32-byte public key to the responding peer and waits for the remote 32-byte public key from the counterparty.
+
+ODD secp256k1 public keys MUST be used (public keys starting with 0x02). If the public key from the generated ephemeral key is an EVEN public key (starting with 0x03), its public key SHOULD be negated and then recalculated.
+Only using ODD public keys makes it more complex to identify the handshake based on analyzing the traffic.
+
+The handshake request and response message are raw 32byte payloads containing no header, length or checksum (the pure 32byte payload) and MUST be sent before anything else.
+
+Public keys starting with the 4-byte network magic are forbidden and MUST lead to local regeneration of an ephemeral-key.
+
+Pseudocode for the ephemeral-key generation
+<pre>
+do {
+    ecdh_key.MakeNewKey();
+    if (ecdh_key.GetPubKey()[0] == 3) {
+        ecdh_key.Negate();
+    }
+} while (m_ecdh_key.GetPubKey()[0..3] == NETWORK_MAGIC);
+</pre>
+
+Once a peer has received the public key from its counterparty, the shared secret MUST be calculated by using secp256k1 ECDH.
+
+Private keys will never be transmitted. The shared secret can only be calculated if an attacker knows at least one private key and the counterparty's public key. This key-exchange is based on the discrete log problem and thus not sufficiently strong against known forms of possible quantum computer algorithms. Adding an additional quantum resistant key exchange like NewHope is possible but out of scope for this proposal.
```

Comment:
is it necessary to use "will" here? Can't one of the capitalized RFC verbs be used?